module/zfs/arc.c

```diff
@@ -2294,9 +2301,9 @@ arc_buf_fill(arc_buf_t *buf, spa_t *spa, uint64_t dsobj, arc_fill_flags_t flags)
 		 */
 		if (arc_buf_try_copy_decompressed_data(buf)) {
 			/* Skip byteswapping and checksumming (already done) */
-#ifdef _KERNEL
-			// XXX #ifdef out of userland because it causes zdb to die
-			ASSERT3P(hdr->b_l1hdr.b_freeze_cksum, !=, NULL);
+#ifdef DEBUG
+			if (zfs_flags & ZFS_DEBUG_MODIFY)
+				ASSERT3P(hdr->b_l1hdr.b_freeze_cksum, !=, NULL);
 #endif
```

Comment:
Ah I see, the other places check ZFS_DEBUG_MODIFY. Guess this could also be what ZOL's https://github.com/zfsonlinux/zfs/issues/5721 needs.

Reply:
Yes, since b_freeze_cksum won't be generated when the ZFS_DEBUG_MODIFY bit is not set, the ASSERT and EQUIV that checks for b_freeze_cksum != NULL should not be done when the ZFS_DEBUG_MODIFY bit is not set either.    

The EQUIV will fail when copied is true but the copied buffer doesn't have a b_freeze_cksum.    

If you just comment out or guard the EQUIV then the ASSERT in the calling function will fail since it is essentially checking the copied-means-frozen_checksum_exists side of the EQUIV.

Since this doesn't blow up in non-DEBUG, this change is safe on its face.   After the tester runs OK, we'll have empirical support for that too.